TRI-227/Create dynamic webhook endpoints

luccar31 opened the pull request:
# Pull Request

## Description
A webhook is generated along with the integration process. This works by implementing the strategy pattern to adapt the service to the issue provider (Linear, Jira, Assana, etc.) selected by the client user, at the moment of requesting the integration with the platform. When finish integrating, a webhook is generated on the external platform. All data needed in stored in the database.

## Ticket Link
[TRI-227](https://linear.app/tricker-sirius/issue/TRI-227/create-dynamic-webhook-endpoints)

luccar31 committed:
feat: add graphql and apollo client dependencies

luccar31 committed:
feat: add apollo client configuration and function to create new instances

luccar31 committed:
feat: add webhook service, implement strategy pattern for external webhook registration and connect webhooks to integrated projects

luccar31 committed:
Merge branch 'TRI-226/Set_webhooks_configuration' into TRI-227/create-dynamic-webhook-endpoints

luccar31 committed:
feat: add default organization administrator user to seed

email must be supplied via environment variables

luccar31 committed:
feat: add loggers to retry function

luccar31 committed:
fix: correct types in IsValidApiKeyConstraint

luccar31 committed:
fix: better error logs with ErrorHandling middleware

luccar31 committed:
fix: transform annotation works correctly now. delete issueProviderName from AuthorizationRequestDTO as it comes in path params

luccar31 committed:
fix: add param issueProviderName to method in integration repository

Refs: #75ec62f

luccar31 committed:
fix: delete unnecessary string transformation

delete IssueProviderURLParams validation from /information endpoint

Refs: #75ec62f

luccar31 committed:
feat: add error handling link to apollo client

luccar31 committed:
fix: change apollo client import to core

luccar31 committed:
fix: correct issueProviderName from lower case to upper case

Refs: #75ec62f

luccar31 committed:
feat: add dynamic webhooks endpoint. it accepts request from all different providers supported

luccar31 committed:
feat: webhook registration url now changes depending on node env.

data return is now correct

add IssueLabel resourceType to linear webhook registration strategy

luccar31 committed:
fix: delete comments.

change param name from issueProviderId to issueNameProvider

webhook sources are now created the right way

luccar31 committed:
fix: add projectId to create project relationships

luccar31 committed:
fix: correct and add ts types and param names

luccar31 committed:
fix: apiKey was not being provided to getAndAdaptProjects

luccar31 committed:
fix: pendingAuths are now deleted when an error occurs in the integration process.

apiKey is now supplied to the methods that needed

issueProviderName is now a parameters to supply in createIntegrationProjectRequest due to changes in AuthorizationRequestDTO

Refs: #75ec62f, #ab8771d, #53f16a7

luccar31 committed:
fix: add LINEAR_GRAPHQL_URL to required env variables

improve ngrok logs

luccar31 committed:
fix: fix double logging

luccar31 committed:
feat: add request params validation

luccar31 committed:
fix: change attribute name to match with request param name

luccar31 committed:
fix: change CustomRequest interface name to a more suitable name

extract logic from rawBodySaver middleware and applied it to all express body parsers

modify imports from CustomRequest with the new name

Juan-Bianchi commented on src/domains/integration/dto/index.ts:
```diff
@@ -221,15 +223,6 @@ export class AuthorizationRequestDTO {
   @IsValidOrganization()
   @IsNotEmpty()
   readonly organizationName!: string;
-
```
Please modify Swagger documentation and let frontend team know they have to modify  /api/integration/:issueProviderName/authorization request body.

Juan-Bianchi commented on src/domains/integration/service/integration.service.impl.ts:
```diff
@@ -66,17 +68,13 @@ export class IntegrationServiceImpl implements IntegrationService {
-      Logger.error(e.cause);
-      Logger.error(e.stack);
       await this.emailService.sendIntegrationErrorEmail(integratorEmail);
-      await this.pendingAuthProjectRepository.update({ pendingProjectId: pendingProject.id, isIntegrating: false });
```
As, delete action is only performed logically, update method to change isIntegrating property should be made as well. It is to have consistent information about pending project. This modification was already made in [TRI-240 pull request](https://github.com/sirius-valley/tricker-backend/pull/108)

Juan-Bianchi commented on src/domains/webhook/service/webhookRegistrationStrategy.ts:
```diff
@@ -0,0 +1,6 @@
+
+export interface WebhookRegistrationStrategy {
+  supports: (issueProviderName: string) => boolean;
+  registerWebhook: (providerApiKey: string, providerProjectId: string, trickerWebhookSourceId: string) => Promise<ProviderWebhookRegistrationData>;
```
You may use destructuring because of the amount of method parameters, and may change the third paramter name to a more generic one.

Juan-Bianchi commented on src/domains/webhook/service/webhook.service.ts:
```diff
@@ -0,0 +1,5 @@
+import { type WebhookSourceDTO } from '@domains/webhookSource';
+
+export interface WebhookService {
+  createWebhook: (apiKey: string, projectId: string, providerProjectId: string, issueProviderName: string) => Promise<WebhookSourceDTO>;
```
Here you may use destructuring too.

Juan-Bianchi commented on src/domains/webhook/service/webhookRegistrationStrategy.provider.ts:
```diff
@@ -0,0 +1,12 @@
+  private readonly strategies: WebhookRegistrationStrategy[] = [new LinearWebhookRegistrationStrategy()];
+
+  getStrategy(issueProviderName: string): WebhookRegistrationStrategy {
+    const strategy = this.strategies.find((strategy) => strategy.supports(issueProviderName));
```
Please add return type

Juan-Bianchi commented on src/domains/webhookSource/dto/index.ts:
```diff
@@ -7,7 +7,7 @@ export class WebhookDynamicRouteParam {
   /** The ID of the webhook. */
   @IsNotEmpty()
   @IsString()
```
Please modify Swagger Documentation

Juan-Bianchi commented on src/domains/webhookSource/dto/index.ts:
```diff
@@ -50,3 +50,16 @@ export class WebhookSourceDTO {
   }
 }
+
+export interface ProviderWebhookRegistrationData {
```
Please add documentation to the interfaces

Juan-Bianchi commented on src/utils/graphql/apollo.ts:
```diff
@@ -0,0 +1,39 @@
+
+const createHttpLink = (apiKey: string): HttpLink => {
+  return new HttpLink({
+    uri: process.env.LINEAR_GRAPHQL_URL!,
```
Please, add LINEAR_GRAPH_URL to .env.template, and if necessary add it to the method that validates all env variables are present

Juan-Bianchi commented on src/domains/webhook/service/webhookRegistrationStrategy.linear.impl.ts:
```diff
@@ -0,0 +1,35 @@
+  async registerWebhook(providerApiKey: string, providerProjectId: string, trickerWebhookSourceId: string): Promise<ProviderWebhookRegistrationData> {
+    const apolloClient = createApolloClient(providerApiKey);
+
+    const url = process.env.NODE_ENV === NodeEnv.DEV ? process.env.NGROK_STATIC_DOMAIN! : 'tricker.com.ar';
```
You may use an env variable for Production url as well

luccar31 replied:
Will be corrected in another PR.

Juan-Bianchi commented on src/domains/webhook/service/webhookRegistrationStrategy.linear.impl.ts:
```diff
@@ -0,0 +1,35 @@
+import { CreateWebhookMutation } from '@utils/graphql/queries';
+import { type WebhookRegistrationStrategy } from '@domains/webhook/service/webhookRegistrationStrategy';
+import { NodeEnv } from '@utils';
+
```
You May put strategy registration files into a separate folder.

luccar31 committed:
feat(adapters): change name from factory to provider, it was not really a factory

add support method to ProjectManagementToolAdapter interface. this method returns true if the adapter support the issue provider

refactor logic from provider to find a suitable adapter

Juan-Bianchi commented on src/domains/webhook/service/webhookRegistrationStrategy.linear.impl.ts:
```diff
@@ -0,0 +1,35 @@
+      mutation: CreateWebhookMutation,
+      variables: {
+        input: {
+          label: `Tricker Project Integration`,
```
You may give a unique identifier to the webhook in linear. This is to avoid problem among developers using dev environment

Juan-Bianchi commented on prisma/seed.ts:
```diff
@@ -46,6 +47,27 @@ async function runSeeding(): Promise<void> {
     update: {},
   });
+
+  const defAdminEmail = process.env.DEFAULT_ADMIN_EMAIL;
```
Please add this variable to .env.template

Juan-Bianchi commented on prisma/seed.ts:
```diff
@@ -46,6 +47,27 @@ async function runSeeding(): Promise<void> {
   });
+
+  const defAdminEmail = process.env.DEFAULT_ADMIN_EMAIL;
+
```
Please add a Logger message for let the developer know organizationAdministrator is being seeded

Juan-Bianchi commented on src/utils/retry.ts:
```diff
@@ -1,19 +1,27 @@
 }
 export const retry = async <ReturnType, ErrorType extends Error = Error>(operation: AsyncOperation<ReturnType>, options: RetryOptions, errorsToStop: Array<new (...args: any[]) => ErrorType>): Promise<ReturnType> => {
   try {
+    Logger.info({ prefix: '<<<retry function>>>', message: `Retrying again... Retries left: ${options.retries}` });
```
You may show this particular message when retrying and not the first time. It may be a bit confusing to see The retrying message in the logger screen.
![image](https://github.com/sirius-valley/tricker-backend/assets/104390122/8800bb62-0792-4653-817f-25f7343db251)

Juan-Bianchi commented on src/domains/integration/service/integration.service.impl.ts:
```diff
@@ -110,11 +108,13 @@ export class IntegrationServiceImpl implements IntegrationService {
 
     Logger.complete(`Integration finished`);
 
+    await this.webhookService.createWebhook(decryptData(pendingProject.token), newProject.id, projectData.projectId, issueProviderName);
```
You may add a logger to inform webhooks creation.

Juan-Bianchi commented on src/domains/webhook/service/webhookRegistrationStrategy.linear.impl.ts:
```diff
@@ -0,0 +1,35 @@
+
+    return {
+      id: mutationResult.webhookCreate.webhook.id,
+      secret: mutationResult.webhookCreate.webhook.secret,
```
Please encrypt secret, validation middleware decrypts it when request is received.
![image](https://github.com/sirius-valley/tricker-backend/assets/104390122/4d63f82d-df5a-4fca-8aec-361cbf674533)

luccar31 committed:
fix: fix integration tests by adding webhookService mock

delete test issueProviderName validation in AuthorizationRequestDTO because class was modified

Juan-Bianchi reviewed (changes requested):
I've left some comments where issues were found

luccar31 committed:
test(webhooks): add tests for webhook service and related classes

luccar31 committed:
fix: secret is now encrypted

luccar31 committed:
fix(docs): delete issueProviderName attr from AuthorizationRequest

Refs: #75ec62f

luccar31 committed:
fix(integration): when pendingAuth is deleted, isIntegrating flag now becomes false

luccar31 committed:
fix: destructuring is now used in some methods

luccar31 committed:
fix: add types to improve readability

luccar31 committed:
fix: webhook endpoint now only admits http post requests

luccar31 committed:
docs: add webhooks endpoint swagger documentation

luccar31 committed:
fix: add required environment vars to .env.template

luccar31 committed:
fix: improve logging when seeding database

luccar31 committed:
fix: improve logging when retry operation

luccar31 committed:
fix: improve logging when registering webhook

luccar31 committed:
fix: move webhook registration related files to their own sub folder

luccar31 committed:
fix: fix malformated doc

luccar31 committed:
fix: fix test due to changes in implementation

Refs: #57a8ee1

luccar31 committed:
merge dev

luccar31 committed:
fix: change name from adapterFactory to AdapterProvider

Refs: #e1f842a

luccar31 committed:
fix: linearMock now implements its interface correctly

some imports are now type imports

luccar31 committed:
fix: retry logger is now disabled when running tests

Juan-Bianchi commented on src/documentation/domains/webhook/controller/webhook.controller.docs.ts:
```diff
@@ -0,0 +1,23 @@
+ *           type: string
+ *           description: The unique identifier of the registered webhook.
+ *       responses:
+ *         '200':
```
As a comment, remember to add in the future the returned Schema and an example of it. In case you decide not to send any DTO you may change your HTTP respoonse number

luccar31 replied:
I decided to not return anything based on the [Linear Docs](https://developers.linear.app/docs/graphql/webhooks#how-does-a-webhook-work). Based on that document, a 200 HTTP code in response to a successful request is enough. Besides, I think you wouldn't want to attach a body to that response, it will not be processed by the client server and you'll be exposing unnecessary data.

Juan-Bianchi commented on src/domains/adapter/factory/adapterProviderImpl.ts:
```diff
@@ -0,0 +1,17 @@
+
+export class AdapterProviderImpl implements AdapterProvider {
+  private readonly adapters: ProjectManagementToolAdapter[] = [linearAdapter];
+
```
Please add documentation to the method. You could use ConflictException too if you want to get the error map by the error handler.

luccar31 replied:
I don't think a 409 status code would be suitable here. As the [docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409) states, it indicates a problem with the _current state of the target resource_. A missing adapter for the provider would not be a conflict with the _current state of the target resource_. It is something you have to fix and redeploy. More over, if the issue provider is not supported, that request should not be even process in the beginning. That's why I just left a 500.

Juan-Bianchi commented on src/utils/graphql/queries.ts:
```diff
@@ -0,0 +1,14 @@
+import { gql } from '@apollo/client/core';
+
+export const CreateWebhookMutation = gql`
```
You could add documentation to the query if you want

Juan-Bianchi commented on src/domains/webhookSource/repository/webhookSource.repository.impl.ts:
```diff
@@ -1,10 +1,24 @@
 export class WebhookSourceRepositoryImpl implements WebhookSourceRepository {
   constructor(private readonly db: PrismaClient) {}
 
+  async create(input: WebhookSourceCreateInput): Promise<WebhookSourceDTO> {
```
Please add documentation to the method

Juan-Bianchi commented on src/domains/webhook/service/webhookRegistration/webhookRegistrationStrategy.linear.impl.ts:
```diff
@@ -0,0 +1,35 @@
+    return issueProviderName === 'LINEAR';
+  }
+
+  async registerWebhook(input: { providerApiKey: string; providerProjectId: string; trickerWebhookSourceId: string }): Promise<ProviderWebhookRegistrationData> {
```
Please add documentation to the method

Juan-Bianchi commented on src/domains/webhook/service/webhook.service.impl.ts:
```diff
@@ -0,0 +1,41 @@
+    private readonly issueProviderRepository: IssueProviderRepository,
+    private readonly strategyProvider: WebhookRegistrationStrategyProvider
+  ) {}
+
```
Please add documentation to the method

Juan-Bianchi reviewed (changes requested):
Please add documentation in all the methods marked

luccar31 committed:
docs(webhook): add js docs to webhook related classes and interfaces
